### ISSUE TITLE
[Proposal]: Let ref structs implement interfaces and substitute into type parameters

### ISSUE THREAD

**Issue by agocke** [dotnet/csharplang#7608]
# Ref structs implementing interfaces

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

There's no particular problem with letting `ref structs` implement interfaces today. The problem is that there's no way to call them. I propose we allow ref structs to substitute type parameters, which would let users call the interface methods safely.

This proposal has two parts. The first part is a constraint, called `box`. This constraint allows boxing the constrained type parameter. It would look something like,

```C#
void M<T>(T t) where T : box
{
   object o = t; // box
}
```

The trick is that this `box` constraint already implicitly exists on all type parameters, even unconstrained type parameters. So we need a way to remove it. The second part of this proposal is a new language feature, called "anti-constraints." An anti-constraint is a constraint that removes, or cancels out, an implicit constraint. We could use the `~` syntax to represent the complement operation which cancels out the implicit constraint. If C# users want to pass ref structs through generic parameters they would now have to declare the parameter as

```C#
void M<T>(T t) where T : ~box
{
   object o = t; // would now produce an error
}
```

With this restriction in place, ref structs would be able to implement interfaces **and** users would be able to call those interface implementations.

## Motivation
[motivation]: #motivation

The motivation here is simple: this is an onerous restriction that is there primarily to ensure ref struct safety. We can ensure ref struct safety with a small new feature that will still let users implement interfaces. Spans should be able to implement IEnumerable, wrappers should be able to implement IDisposable, etc.

The current restriction is a side effect of a rule to preserve `ref struct` safety. Ref structs can never appear in the heap, meaning that they can never be boxed. To invoke an interface implementation on a struct in C#, the user must either box the struct instance to the interface and then call the method, or pass the instance to a method with a generic parameter that is constrained to the interface. Ref structs cannot be boxed, and therefore the first option is never viable. The second option is not _currently_ viable because C# does not place any restrictions on boxing generic parameters inside generic methods. This is easily fixable.

## Detailed design
[design]: #detailed-design

We don't need much more than the above. We would have to go through and ensure that ref struct boxing is prohibited in all the right places, but this is already implemented for boxing to `object`.

## Drawbacks
[drawbacks]: #drawbacks

Anti-constraints are a new language feature. But not a complicated one, and one that we could likely reuse in the future if we ran into similar problems in other domains.

## Alternatives
[alternatives]: #alternatives

The primary alternative is a `where T : ref struct` constraint, which is the wrong way around. You don't want to require that the type parameter be a ref struct, you just want to prevent boxing.

## Unresolved questions
[unresolved]: #unresolved-questions

None.

## Design meetings

TBD

**Comment by agocke** [comment]
@MadsTorgersen @333fred Could we get this on a triage list? It feels like we've spent more hours in LDM working around this problem than we would actually fixing it.

**Comment by huoyaoyuan** [comment]
Some comments not directly-related, but interesting to consider:

What about `void` as a generic parameter? It's more complicated in type system, but there's not safety problem. Variables or fields declared as `void` can be omitted. Comparing with `ref struct`, the omitted `void` can appear in non ref-struct types.

The unconstraint can solve the gap for concrete methods, but how about delegate types? The delegate type parameters can't be constrained, should we specialize for them?

**Comment by agocke** [comment]
Also, @jaredpar  for lifetime issues. My thinking was that every parameter would implicitly carry a "heap-safe-to-escape" lifetime, but maybe there's something I'm not thinking of. Obviously people will request more lifetime stuff in the future to workaround that, but I figured that could be an additional set of language features that we do later.

**Comment by En3Tho** [comment]
Can you clarify an example with `where T : ~box`: how does it affect boxable types like classes? It feels to me like `~box` allows ref structs only?

**Comment by jaredpar** [comment]
Rather than `where` I think we should consider `allow`. There are more anti-constraints that have been proposed for the language including items like allowing pointers. Having a new keyword then a `~` for them feels unnatural where as the following feels much more natural to me: 

```csharp
void M<T>() 
  where T : IDisposable
  allow T : ref struct { 

}
```

> We don't need much more than the above.

Disagree. As a `ref struct` can satisfy the `where T : ~box` proposal. That means that we have to consider all `T` which satisfy `~box` to be potentially `ref struct`. The proposal doesn't seem to do this but rather focuses on the boxing aspect. That means without additional restrictions code like the following would be allowed: 

```csharp
class C<T> where T : ~box {
  T _field;
}

C<Span<int>> // not good
```

That is also why I prefer `allow T : ref struct` because it's very clear this is the capability being allowed here and would make it much clearer why the above code fails to compile (also makes it easier for us to generate better error messages).

**Comment by agocke** [comment]
> No. You need to prevent all behaviors that are disallowed by a ref struct. Boxing is only one of them.

Yeah I agree with that. And I think you're right that we might want to use different syntax. But I think the basic concept is still the same here.

**Comment by jaredpar** [comment]
There are a couple of other items. 

Essentially any `T` constrained to `allow T : ref struct` needs to have all capabilities and restrictions of a `ref struct`: 

1. Cannot be a field of a non `ref struct` 
2. Declarations can be marked `scoped` 
3. Is subject to lifetime tracking rules as a `ref struct` value would be
4. etc ... 

Effectively for all intent and purposes the compiler would treat it like a `ref struct` value. Cause it potentially can be and that is the _worst case_ from a restriction point of view. No harm is done to other types by being given such restrictions. 

Another other item that's come up here is how such an anti-constraint should be applied to existing APIs in the core libraries. Consider for example that every variation of `Func / Action` should have `allow T : ref struct` for all their type parameters. In those cases it's a 100% upside change. 

One approach to this is to simply go through and manually update every such `delegate` definition to have this anti-constraint. That's manual and benefits only the developers who put in the work. Another option is to simply embed it in the language rules. Essentially for a type parameter on a `delegate` definition where it only appears only as the type of a simple parameter  (no `in/out/ref` modifiers) the language will automatically infer `allow T : ref struct`. Doing so would make the `delegate` more expressive than if it didn't. I've gone back and forth on this but generally lean towards automatic inferring it here. Could be talked off this point though.

Another item is that allowing `ref struct` to implement interfaces means we should consider allowing `[UnscopedRef]` attributes on `interface` members. Lacking that attribute there are patterns expressible in `ref struct` today that cannot be expressed when factored into an `interface`. Think that is relatively straight forward (has meaning when implemented on a `struct` but not on a `class`) but still needs to be spec'd out.

**Comment by davidwrighton** [comment]
On top of all of those details that @jaredpar brings up, I'd really like to be able to use this anti-constraint with `Span<T>` so that we can do a better job around dealing with certain InlineArray scenarios, but that api is just messy.

**Comment by TahirAhmadov** [comment]
I think the notion of anti-constraints needs to be considered very thoroughly. It's almost a full on massive change all on its own. What do `~class`, `~struct`, `~new`, etc. mean? `~notnull` becomes weird, because it's a double negative. What about types (classes and interfaces)? I think I saw that brought up before, but I don't remember the discussion and why it was desirable.

The proposed `allow` keyword is also a little finicky. Does `allow` only allow (pun intended) `ref struct`? Are there any potential other scenarios? Why not just add `ref struct` to the list of `where` - surely adding a new keyword here is overkill?

**Comment by HaloFour** [comment]
> I think the notion of anti-constraints needs to be considered very thoroughly

I don't think they're intended to be a general purpose feature.  I think they're only to be used where they enable the language to support additional functionality.  But they also don't make sense as a normal constraint.  A `ref struct` constraint doesn't force the generic type argument to be a `ref struct`, it forces the code of the method to follow the rules as if it could be a `ref struct`, but the argument could be a normal `struct` (or even a `class`?)

**Comment by jaredpar** [comment]
Put up a full proposal for this feature 

https://github.com/dotnet/csharplang/blob/main/proposals/ref-struct-interfaces.md

**Comment by timcassell** [comment]
Another alternative that was mentioned in the old issue is placing the `ref` keyword at the generic declaration site, much like `in` and `out` on co/contravariant interfaces.

```cs
public void M<ref T>(T o) where T : IDisposable
    => o.Dispose();
```

**Comment by CyrusNajmabadi** [comment]
My main issue there @timcassell is that reads to me like "this must be a ref struct" as opposed to "this is allowed to be a ref-struct".

**Comment by TahirAhmadov** [comment]
What about `where T : allow ref struct, ISomeInterface`? It's kind of strange to have 2 places with constraints and "expanders" for one generic argument.

**Comment by CyrusNajmabadi** [comment]
The syntax here is the least interesting part :)   We'll likely consider a bunch and settle on one the group feels conveys the idea the best and feels the most c#-y :)

**Comment by jaredpar** [comment]
@timcassell 

> Another alternative that was mentioned in the old issue is placing the ref keyword at the generic declaration site, much like in and out on co/contravariant interfaces.

For me that is much too easily confused with supporting the ability to have `ref` as a type modifier. For example `List<ref int>`. That is a much different feature (and not one that I think anyone is considering given the fundamental change it would require to the runtime). 

> What about where T : allow ref struct, ISomeInterface? It's kind of strange to have 2 places with constraints and "expanders" for one generic argument.

I agree it's a bit strange to have both `allow` and `where` but unfortunately there need to be up and down modifiers here. That is because a generic parameter today has implicit constraints (can't be pointer, `ref struct`, etc ...). The syntax must provide a method to remove the implicit constraints (t`allow`) and add more (`where`). 

As @CyrusNajmabadi mentioned though the syntax is the least interesting part. It is also likely the part we will end up debating the most in LDM. For now I would encourage people to focus on the behaviors and ensure they satisfy the scenarios you want to get out of this feature.

**Comment by TahirAhmadov** [comment]
The funny thing that having read the proposal and thought about it, it kind of makes sense to me so the only remaining question that I have is the syntax :) It's just not "what looks nice", but potential future enhancements. I think we should all imagine what `allow` can allow (pun intended) in the future, if anything; that IMO should be the main factor to decide the syntax.

**Comment by timcassell** [comment]
How will calling methods on `System.Object` like `ToString()` and `GetHashCode()` work? From what I understand, if a struct overrides those methods, they will not be boxed when called. But if they don't override them, they will be boxed. Will calling those methods simply be disallowed? Or that would be allowed in this case since the box is short-lived and doesn't escape?

**Comment by En3Tho** [comment]
@TahirAhmadov I thought the same. But I don't really like "allow" syntax. I'd rather think  of this not as "allow" but as "restrict the lifetime of values of certain type T".

And I would like to see some work in this direction. E.g. if I didn't want some IDisposable like native resource handle to escape I would use similar mechanism.

This can lay groundwork for more general lifetime and ownership features. I think it's important to think of them when designing syntax/rules.

**Comment by IS4Code** [comment]
While the idea to have `<ref T>` instead of `allow` came from me, I think ultimately `allow` is best and future-consistent approach. In the end we could have `allow T : static`, `allow T : *` (pointers), `allow T : ref` (`ref` to anything) etc. if needed, so this syntax could prepare the ground for it.

**Comment by hez2010** [comment]
While this also allow us to use `allow` pattern outside the `ref struct` scenarios?

For example,

```cs
T ParseValue<T>(string str) where T : IParseable<T> allow T : string
{
    if (typeof(T) == typeof(string))
    {
        return (T)str;
    }
    return T.Parse(str);
}
```

**Comment by HaloFour** [comment]
@hez2010 

> Will this also allow us to use `allow` pattern outside the `ref struct` scenarios?

Not without said features being explicitly designed and implemented.

**Comment by jhudsoncedaron** [comment]
So I just got news of this proposal. There's a major problem here: "adding instance default interface methods to existing interfaces becomes universally source+binary breaking". We depend on the ability to add default interface methods all over the place.

I'm reading this and going "really?" This breaks the primary motivating feature to have default interface methods.

I did come to an idea of how to fix it but this is probably bonkers hard: when you encounter a missing default interface method on a ref struct; jit the mehod and bail if the jit would emit a box operation.

**Comment by jaredpar** [comment]
>  This breaks the primary motivating feature to have default interface methods.

Default interface methods have always had edge cases where they do not work. For example the diamond problem where the runtime cannot pick the best DIM member results in runtime exceptions. This is just another case where DIM will have an edge case. 

>  when you encounter a missing default interface method on a ref struct; jit the mehod and bail if the jit would emit a box operation.

I suspect that would not be very effective in practice. Most of the DIM members I've seen rely on calling other available members in the type. That implicitly uses `this` which forces a boxing operation on the value. This would really only work for DIM members that call into `static` members only and do not use `this`. Not sure that is a large amount.

**Comment by jhudsoncedaron** [comment]
"That implicitly uses this which forces a boxing operation on the value.": My testing is that doesn't box a struct. The jitter is smarter than that.

**Comment by jhudsoncedaron** [comment]
How does that not break default implementations on normal mutable structs? (As in, the default implementation generates nonfunctional code because it mutates a copy)?

**Comment by reflectronic** [comment]
It does break them: 
```cs
using System;

var s = new S();
Mut(ref s);

// Outputs '0'
Console.WriteLine(s.V);

static void Mut<T>(ref T t) where T : IMutate
{
    t.Mutate();
}


interface IMutate
{
    int V { get; set; }
    void Mutate() { V++; }
    
}

struct S : IMutate
{
    public int V { get; set; }
    // Uncomment to print '1' instead
    // public void Mutate() { V++; }
}
```

https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQ6BuBDMACAZzwF48oBTAdzwGUAKASgG50BZCYOscgM0OfToA9ELwB5DgAcORAOQAGWejhIAnHQIA6AGoCMaFQDY8cACx52wADwAVAHxdeeG3mAM8lABbluzvCDwASUtsYHJ0AG90PBjXTRCwxhY0AF9BfQBLKDCwHmwAY3IghPC0KLRYvCzgPG08CLwAc3JgJkIWtrSK2LMLDlDyRnragGoRzujY9C70AmBIfJqaf2L+sMjJmLgAZirs2uHm1vbjrsqRPABVKHyAewBbe/J94Fu8STBqvFkkWT258jYAAmmzwFx2JnMJSGDW0YwmqXQQA==

**Comment by huoyaoyuan** [comment]
How to solve "`Tuple<T1, T2>` should be ref struct when either of T1 or T2 is struct"?
It may require some specification work. Will a new `RefTuple<T1, T2>` type be added to workaround this?

**Comment by ds5678** [comment]
I have a proposal to solve the `Span<Span<T>>` issue.

> In this comment, "unusable type" refers to any invalid type construction involving ByRefLike types.

* We permit method parameters, method returns, and unbacked properties to possibly have an unusable type, perhaps with a warning message.
* Expressions (such as `default`) and local variables are not permitted to possibly be an unusable type.
* We introduce pattern matching for type parameters, perhaps limited to the use case `T is ref struct`.

```cs
readonly ref struct Span<T>
{
    public Span(T[] array)//Warning: parameter has a potentially unusable type
    {
        if (T is ref struct)
        {
            throw new NotSupportedException();
        }
        else
        {
            //Current implementation
        }
    }
}

//Error: expressions cannot target a potentially unusable type
_ = new Span<Span<T>>(default(Span<T>[])).Length;

static void Foo<T>() allow T : ref struct
{
    //Error: local variables cannot target a potentially unusable type
    T[] bar;
}
```

**Comment by jhudsoncedaron** [comment]
I've sat and considered for two months the default interfaces on ref structs thing. I've come to the conclusion my initial idea was more correct than anything else. Since ref structs are a new thing there's no reason a priori we can't change the rule so that default interface methods on ref structs get the actual ref struct and not a copy, so it's only going to fail if the jitter can't remove the implicit boxing given the il code of the default member. This will definitely be the case if it passes it to something expecting object, but most likely won't be the case if it accesses other interface members. Since you have the type of the interface, every such call gives the jitter more type information for locals. Note that such jitter code is far beyond my capacity to write.

**Comment by jaredpar** [comment]
> Since ref structs are a new thing there's no reason a priori we can't change the rule so that default interface methods on ref structs get the actual ref struct and not a copy, so it's only going to fail if the jitter can't remove the implicit boxing given the il code of the default member

This sounds like a discussion to have with the runtime team. Given that they told us this wasn't a solvable problem last time though I'm going on that information.

**Comment by OJacot-Descombes** [comment]
Wouldn’t it be possible to do a stackalloc-kind of boxing? This would allow ref structs to be boxed (and it would also allow optimization of boxing in many other cases).

**Comment by huoyaoyuan** [comment]
> Wouldn’t it be possible to do a stackalloc-kind of boxing? This would allow ref structs to be boxed

This won't solve the problem. With escape analysis, objects with limited lifetime can be moved onto stack. However, current generic parameter allows unlimited lifetime of defined variables. We need to ensure that ref struct isn't stored into any long-lived position.

**Comment by jhudsoncedaron** [comment]
It's almost like ref struct : IInterface wants to be a new kind of interface. Taking existing generics that have constraint : IInterface won't work then.

**Comment by timcassell** [comment]
> > Wouldn’t it be possible to do a stackalloc-kind of boxing? This would allow ref structs to be boxed
> 
> This won't solve the problem. With escape analysis, objects with limited lifetime can be moved onto stack. However, current generic parameter allows unlimited lifetime of defined variables. We need to ensure that ref struct isn't stored into any long-lived position.

It wouldn't work in general, but I think it would work for this case to allow calling `Object` methods that I asked about earlier (`ToString`, `GetHashCode`).

**Comment by TahirAhmadov** [comment]
Question: would something like below become possible?
```cs
Span<char> span = ...;
string str = string.Create(10, span, (destSpan, sourceSpan)=> { ... });
// or
string str = string.Create(10, (x, y, span), (destSpan, state)=> { ... });
```

**Comment by jaredpar** [comment]
Assuming all of the delegates are changed to have the new anti constraint and there is no capture it should be possible.

**Comment by timcassell** [comment]
Just curious, because I doubt anyone uses it anymore, but how would delegate `BeginInvoke` work with ref structs?

**Comment by jnm2** [comment]
@timcassell BeginInvoke throws PlatformNotSupportedException on newer .NET runtimes since .NET Core, and it sounds like the feature being discussed here would require runtime support and thus exclusively be for future .NET runtimes.

**Comment by timcassell** [comment]
I guess it was already possible by declaring a delegate with non-generic ref struct, so it's a non-issue for this feature.

**Comment by jaredpar** [comment]
> In theory I'm not seeing why we couldn't update Func and Action and simplify this whole API surface.

My expectation is that these core delegates are all updated to have `allows ref struct`. The [proposal][proposal] actually speculated as to whether we should just automatically do this but ended up deciding against it. But expectation is that core types like `Func`, `Action`, `IEnumerable` would move to adopt this new anti-constraint. 


[proposal]: https://github.com/dotnet/csharplang/blob/main/proposals/ref-struct-interfaces.md#auto-applying-to-delegate-members

**Comment by jaredpar** [comment]
> Would F# need to suppress its default automatic generalization and generic constraint propagation for this specific (anti-)constraint? That would be a pretty major departure for F#.

Think that is a question for F# language designers. C# behavior here is essentially following how constraints are modeled in IL. There is nothing stopping F# from providing a different presentation here. 

@vzarytovskii

**Comment by brianrourkeboll** [comment]
> There is nothing stopping F# from providing a different presentation here.

Yes, but my comment was in part about how the choices that the current design makes available to F# here don't seem ideal:

- Treat this constraint differently from the way F# treats all others.
- Deal with the implications of this constraint being propagated through type signatures everywhere.
- A secret, third thing?

But it's entirely possible that I'm missing something or haven't thought things through enough, and that consuming this from F# will be simpler than I'm making it out to be.

**Comment by HaloFour** [comment]
> Treat this constraint differently from the way F# treats all others.

Given anti-constraints are different from other constraints, it would feel very appropriate if F# would treat them differently than it treats constraints today.

**Comment by jaredpar** [comment]
> Treat this constraint differently from the way F# treats all others.

This is how C# treats constraints today though: they do not propagate by default. This decision is following our existing patterns.

**Comment by brianrourkeboll** [comment]
> > Treat this constraint differently from the way F# treats all others.
> 
> Given anti-constraints are different from other constraints, it would feel very appropriate if F# would treat them differently than it treats constraints today.

That's fair. In a way, adding an "anti-constraint" is really just adding a case to the implicit default constraint union `T` ≈ `where T : default` ≈ `where T : (class | struct)`:

If —

```csharp
void M<T>(T t) { … }
```

≈

```csharp
void M<T>(T t) where T : default { … }
```

≈

```csharp
void M<T>(T t) where T : (class | struct) { … }
```

— then

```csharp
void M<T>(T t) where T : allows ref struct { … }
```

≈

```csharp
void M<T>(T t) where T : (class | struct | ref struct) { … }
```

≈

```csharp
void M<T>(T t) where T : (default | ref struct) { … }
```

🙃🙃🙃

**Comment by agocke** [comment]
Not really. The anti-constraint is about cancelling out a constraint that already exists. The way it currently works is that .NET has an implicit `heap` constraint that's available to all unconstrainted type parameters. That constraint allows you to do things like box variables of that type, and store variables of that type as fields in classes and structs.

The "allows ref struct" anti-constraint cancels out the implicit `heap` constraint and removes it from the constraint list. It's not about adding constraints to an implicit union, it's about removing them.

**Comment by brianrourkeboll** [comment]
I think we're using the same metaphor to refer to different things 

> The \"allows ref struct\" anti-constraint cancels out the implicit heap constraint and removes it from the constraint list. It's not about adding constraints to an implicit union, it's about removing them.

Yeah, that's what my imaginary "constraint union" syntax was supposed to mean — `class` constraint (and all that implies , including ability to be boxed on the heap) _or_ `struct` constraint (and all that implies , including ability to be boxed on the heap) _or_ `ref struct` constraint (and all that implies , including inability to be boxed on the heap). Any concrete type parameter can and must satisfy exactly one of those at a time, i.e., it is either a class, a regular struct, or a ref struct.

This is in contrast to constraint intersections like `where T : class, IFace`.

Incidentally, F# already does represent constraint intersections with `and`: `'T when 'T : not struct and 'T :> IFace`.

**Comment by brianrourkeboll** [comment]
> The \"allows ref struct\" anti-constraint cancels out the implicit heap constraint and removes it from the constraint list. It's not about adding constraints to an implicit union, it's about removing them.

It is indeed an intersection, not a union...

Just like adding `notnull`, I suppose, except `notboxable`...

**Comment by brianrourkeboll** [comment]
...Upon further thought, though, I think I'm back to where I was before. (Sorry for messing up the threading; I was on my phone.)

I think we _were_ using the same metaphor to refer to different things, or rather referring to the same thing from different perspectives (API consumer versus API implementer).

Given my previous post:

> In a way, adding an "anti-constraint" is really just adding a case to the implicit default constraint union `T` ≈ `where T : default` ≈ `where T : (class | struct)`:
> 
> If —
> 
> ```csharp
> void M<T>(T t) { … }
> ```
> 
> ≈
> 
> ```csharp
> void M<T>(T t) where T : default { … }
> ```
> 
> ≈
> 
> ```csharp
> void M<T>(T t) where T : (class | struct) { … }
> ```
> 
> — then
> 
> ```csharp
> void M<T>(T t) where T : allows ref struct { … }
> ```
> 
> ≈
> 
> ```csharp
> void M<T>(T t) where T : (class | struct | ref struct) { … }
> ```
> 
> ≈
> 
> ```csharp
> void M<T>(T t) where T : (default | ref struct) { … }
> ```

And your response:

> The "allows ref struct" anti-constraint cancels out the implicit `heap` constraint and removes it from the constraint list. It's not about adding constraints to an implicit union, it's about removing them.

This emerges from the fact that the set of operations that can be applied to a union are the _intersection_ of the operations that can be applied to every case in the union.

From my other post (again, sorry for messing up the thread):

> Yeah, that's what my imaginary "constraint union" syntax was supposed to mean — `class` constraint (and all that implies, including ability to be boxed on the heap) _or_ `struct` constraint (and all that implies, including ability to be boxed on the heap) _or_ `ref struct` constraint (and all that implies, including inability to be boxed on the heap). Any concrete type parameter can and must satisfy exactly one of those at a time, i.e., it is either a class, a regular struct, or a ref struct.

But by definition in a generic construct the type parameter is not yet concrete, so, since you can't box a `ref struct`, you can't box a `T` `where T : (class | struct | ref struct)`, since `T` might end up being a `ref struct`.

Seen that way, it actually makes sense that F# might _not_ auto-propagate this constraint unless explicitly specified, since, unlike all existing constraints utterable in C# or F#, it is not being _intersected_ with the implicit `(class | struct)` union of constraints.

Given two functions like this (imaginary syntax; F# does happen to already use `and` to represent constraint intersection, so it would probably make sense to use `or` to represent constraint union):

```fsharp
let f (x : 'T when 'T : struct or 'T : not struct or 'T : byref<struct>) = ignore x
let g x = f x
```

It would make sense that `g` would have the _more general_ signature:

```fsi
val f : x:'T -> unit
```

i.e., with the default, implicit constraint union

```fsi
val f : x:'T -> unit when 'T : struct or 'T : not struct
```

since `'T : struct or 'T : not struct` is of course a subset of `'T : struct or 'T : not struct or 'T : byref<struct>`.

Just because `f` has placed the restriction upon itself that `T` might be something unboxable doesn't mean that `g` must also have that restriction.

**Comment by jaredpar** [comment]
> Would just the implementation of this proposal allow the compiler to synthesize the record members or would that be a separate proposal? 

That would need to be a separate proposal. The inability to implement `IEquatable<T>` is just one of the issues that would need to be addressed before we could do `ref record struct`. Guessing when we sit down and look at all of the behaviors around `record` we'd find a few more we'd need to think about.

**Comment by timcassell** [comment]
> One of the marquee original intentions for `ref struct`s implementing interfaces was for `Span<T>` and `ReadOnlySpan<T>` to implement `IEnumerable<T>`. This would help
  solve a number of betterness issues with adding new APIs, but because `IEnumerable<T>.GetEnumerator()` returns an `IEnumerator<T>`, we can't implement it in an allocation-free
  manner. That would cause any `IEnumerable` API to become a performance trap for `Span<T>`, which is extremely undesirable.

Just add the `IEnumerator<T>` interface on the `Span<T>.Enumerator` type. I don't need the `IEnumerable<T>` interface on the span itself for now. Anyway, that could be saved for a future improved `IEnumerable<T, TEnumerator>` interface after C# improves generic inference for those.

My use case that I was hoping to start using this for:

```cs
public static Promise All(params ReadOnlySpan<Promise> promises)
    => All(promises.GetEnumerator());

public static Promise All<TEnumerator>(TEnumerator promises) where TEnumerator : IEnumerator<Promise>, allows ref struct
{
    using (promises)
    {
        ...
        while (promises.MoveNext())
        {
            var p = promises.Current;
            ...
        }
        return ...;
    }
}
```

> Rules for `ref struct`s in generics are approved. Rules for `ref struct`s implementing interfaces generally look good, but need validation against real world scenarios before we
allow this part of the feature to ship in anything more than preview.

Allowing ref struct as generics but disallowing them to implement interfaces kind of neuters the feature.

**Comment by ds5678** [comment]
> Allowing ref struct as generics but disallowing them to implement interfaces kind of neuters the feature.

While I agree that having this held back behind the preview language feature is undesirable, it's still available to the people who really want to use it. For example, I rewrote my HTML generation library to make use of the features.

https://github.com/AssetRipper/AssetRipper.Text.Html
